### PR TITLE
feat: add support for exceptions to annotations

### PR DESCRIPTION
**Pull request by jackhodgkiss**
Some secrets when redacted are rejected due to invalid characters. One example of this would `prometheus_bcrypt_salt` which is rejected due to the use of underscores. Therefore the solution is to add support for exceptions to the standard annotation format.

See: https://github.com/ansible/ansible/blob/devel/lib/ansible/utils/encrypt.py#L118C46-L118C46
https://github.com/openstack/kolla-ansible/blob/master/ansible/roles/prometheus/templates/prometheus-web.yml.j2#L3

### CODE REVIEW

**Review by MaxBed4d — APPROVED** [review]
Discussed in Google meeting.

File: utils/kayobe-automation-redact [hunk 1]
```diff
@@ -5,11 +5,21 @@ import hashlib
...
+}
+
 def annotate(ctx, value):
-  if not isinstance(value, str):
-    return value
-  path_str = map(str, ctx['path'])
-  return f"{'.'.join(path_str)}.{ value }"
+    if not isinstance(value, str):
+        return value
+    path_str = ctx['path'][0]
```
Comment (markgoddard) [inline utils/kayobe-automation-redact:15]: I'm not exactly sure what the path is here, but we've lost anything after the first part of it here, whereas previously it was joining with `.`.

Comment (jackhodgkiss) [reply utils/kayobe-automation-redact:15]: ctx['path'] contains a list with a single element which is the line from file being redacted.

Something like this.

```py
['prometheus_bcrypt_salt']
```

My change simply extracts the first and only element and does away with the `map` where `.join` was being used to expand the `map` object.

Original approach

```yaml
---
bar_baz: bar_baz.original
foo_bar: foo_bar.original
prometheus_bcrypt_salt: prometheus_bcrypt_salt.original
```

Fixed approach for bcrypt_salt

```yaml
---
bar_baz: bar_baz.original
foo_bar: foo_bar.original
prometheus_bcrypt_salt: prometheusbcryptsalt.o
```

Comment (markgoddard) [reply utils/kayobe-automation-redact:15]: Are you sure that there will never be other items in the path list? It looks like @jovial at least considered that case. I'll defer to him.

Comment (jovial) [reply utils/kayobe-automation-redact:15]: I thought the ssh keys were nested dicts:
```
bifrost_ssh_key:
  private_key: value
  public_key: value
 ```
 so path be something like:
 ```
 ['bifrost_ssh_key', 'private_key']
 ```
 just makes it a bit ambiguous where it came from if you only use `private_key`

Comment (jovial) [reply utils/kayobe-automation-redact:15]: annotation_exceptions seems like a reasonable approach though. Could keep the existing behaviour for the non-exception case?

**Comment by jackhodgkiss** [comment]
Most recent change ensures that nested dicts are redacted and still keep the full name.

```
bifrost_ssh_key:
  private_key: value
  public_key: value
```

becomes

```
bifrost_ssh_key:
  private_key: bifrost_ssh_key_private_key.original
  public_key: bifrost_ssh_key_public_key.original
```

**Review by jovial — APPROVED** [review]

**Comment by jovial** [comment]
LGTM - nice solution :)

**Comment by technowhizz** [comment]
@jackhodgkiss Is planned to merge soon?

**Comment by jackhodgkiss** [comment]
> @jackhodgkiss Is this planned to merge soon?

Happy to merge it now. It was working last time I checked but don't have time this week to test again. However, if you encounter any issues please let me know.

**Pull request merged**